### PR TITLE
Forbid using `float("NaN")`.

### LINKED ISSUE #1383: Forbid `float('NaN')`
# Rule request

## Thesis

Forbid `float('NaN')` usage. It's not about forbidding `NaN` at all because it can't be detected without a proper type checking. `math.nan` is still allowed.

## Reasoning

NaN is really confusing and is a good way to catch a lot of unexpected bugs:

```
>>> sorted([5.0, math.nan, 10.0, 0.0])
[5.0, nan, 0.0, 10.0]

>>> 3 < math.nan
False
>>> 3 > math.nan
False

# no commutativity
>>> min(3, math.nan)
3
>>> min(math.nan, 3)
nan
```

Even if you're 100% sure what you're doing, use `math.nan` instead. Forbidding `nan` in any form (maybe, even including numpy) can be done later and is outside of this rule request.

### PR DESCRIPTION
**Pull request by lensvol**
# I have made things!

## Checklist

- [X] I have double checked that there are no unrelated changes in this pull request (old patches, accidental config files, etc)
- [X] I have created at least one test case for the changes I have made
- [X] I have updated the documentation for the changes I have made
- [X] I have added my changes to the `CHANGELOG.md`

## Related issues

Closes #1383 

🙏 Please, if you or your company is finding wemake-python-styleguide valuable, help us sustain the project by sponsoring it transparently on https://opencollective.com/wemake-python-styleguide. As a thank you, your profile/company logo will be added to our main README which receives hundreds of unique visitors per day.

### CODE REVIEW

File: tests/test_visitors/test_ast/test_functions/test_floating_nan.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+@pytest.mark.parametrize('prefix', [
+    'b',
+    'u',
+    '',
+])
+@pytest.mark.parametrize('nan_variant', [
+    'NaN',
+    'NAN',
+    'nan',
+    'Nan',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_functions/test_floating_nan.py:42]: What about `NAn` and `nAn`? Just asking.

File: wemake_python_styleguide/violations/best_practices.py [hunk 2]
```diff
@@ -2194,3 +2196,31 @@ class NonTrivialExceptViolation(ASTViolation):
...
+class FloatingNanViolation(ASTViolation):
+    """
+    Forbids to use ``float("NaN")`` construct to generate NaN.
+
+    Reasoning:
+        This method to generate NaN is really confusing and is a good way to
+        catch a lot of unexpected bugs.
+
+    Solution:
+        Even if you're 100% sure what you're doing, use math.nan instead.
```
Comment (sobolevn) [inline wemake_python_styleguide/violations/best_practices.py:2211]:
```
``math.nan``
```

File: tests/test_visitors/test_ast/test_functions/test_floating_nan.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from wemake_python_styleguide.visitors.ast.functions import (
+    FloatingNanCallVisitor,
+)
+
+
+@pytest.mark.parametrize('code', [
+    'float(4.5) + 4.2',
+    'float(2) + 0',
+    'result = "NaN".join(["a", "a"])',
+    'a = math.nan',
```
Comment (sobolevn) [inline tests/test_visitors/test_ast/test_functions/test_floating_nan.py:15]: Let's add `my_function('nan')` case. It should be allowed.

File: wemake_python_styleguide/visitors/ast/functions.py [hunk 4]
```diff
@@ -172,6 +173,26 @@ def _is_call_ignored(self, node: ast.Call) -> bool:
...
+        self.generic_visit(node)
+
+    def _check_floating_nan(self, node: ast.Call) -> None:
+        if len(node.args) != 1:
+            return
+
+        if not isinstance(node.args[0], (ast.Str, ast.Bytes)):
+            return
+
+        if node.args[0].s.lower() in frozenset(('nan', b'nan')):
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/functions.py:192]: let's move this to a class-level constant!

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thanks a lot! 👍 

Looks like currently you have covered all `'nan'` args, not just to `float` function.

**Comment by lensvol** [comment]
There, I fixed it!

**Review by orsinium — APPROVED** [review]
Great work! The tests are fun, I've never seen before more than one `pytest.mark.parametrize` per test function :)

File: wemake_python_styleguide/visitors/ast/functions.py [hunk 5]
```diff
@@ -189,7 +191,10 @@ def _check_floating_nan(self, node: ast.Call) -> None:
         if not isinstance(node.args[0], (ast.Str, ast.Bytes)):
             return
 
-        if node.args[0].s.lower() in frozenset(('nan', b'nan')):
+        if not isinstance(node.func, ast.Name) or node.func.id != 'float':
```
Comment (sobolevn) [inline wemake_python_styleguide/visitors/ast/functions.py:194]: We have https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/logic/tree/functions.py#L13 just for this case.

**Review by sobolevn — CHANGES_REQUESTED** [review]
Thanks! 🚀

**Review by sobolevn — APPROVED** [review]
Great! 🥇

**Pull request merged**